diff --git a/sound/soc/qcom/sm8150.c b/sound/soc/qcom/sm8150.c
--- a/sound/soc/qcom/sm8150.c
+++ b/sound/soc/qcom/sm8150.c
@@ -297,8 +297,8 @@
 
 static int sm8150_snd_startup(struct snd_pcm_substream *substream)
 {
-	unsigned int fmt = SND_SOC_DAIFMT_CBS_CFS;
-	unsigned int codec_dai_fmt = SND_SOC_DAIFMT_CBS_CFS;
+	unsigned int fmt = SND_SOC_DAIFMT_CBC_CFC;
+	unsigned int codec_dai_fmt = SND_SOC_DAIFMT_CBC_CFC;
 	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
 	struct snd_soc_card *card = rtd->card;
 	struct sm8150_snd_data *data = snd_soc_card_get_drvdata(card);
